lms/djangoapps/mobile_api/tests.py

```diff
@@ -21,35 +26,233 @@
 )
 
 
-@ddt.ddt
-class TestMobileApiUtils(ModuleStoreTestCase, APITestCase):
+class MobileAPITestCase(ModuleStoreTestCase, APITestCase):
     """
-    Tests for mobile API utilities
+    Base class for testing Mobile APIs.
+    Subclasses are expected to define REVERSE_NAME to be used for django reverse URL.
+    They may also override any of the methods defined in this class to control the behavior of the TestMixins.
     """
-
     def setUp(self):
+        super(MobileAPITestCase, self).setUp()
+        self.course = CourseFactory.create(mobile_available=True)
         self.user = UserFactory.create()
+        self.password = 'test'
+        self.username = self.user.username
+
+    def tearDown(self):
+        super(MobileAPITestCase, self).tearDown()
+        self.logout()
+
+    def login(self):
+        """Login test user."""
+        self.client.login(username=self.username, password=self.password)
+
+    def logout(self):
+        """Login test user."""
+        self.client.logout()
+
+    def enroll(self, course_id=None):
+        """Enroll test user in test course."""
+        CourseEnrollment.enroll(self.user, course_id or self.course.id)
+
+    def unenroll(self, course_id=None):
+        """Unenroll test user in test course."""
+        CourseEnrollment.unenroll(self.user, course_id or self.course.id)
+
+    def login_and_enroll(self, course_id=None):
+        """Shortcut for both login and enrollment of the user."""
+        self.login()
+        self.enroll(course_id)
+
+    def get_response(self, reverse_args=None, expected_response_code=200, **kwargs):
+        """
+        Helper method for calling endpoint, verifying and returning response.
+        If expected_response_code is None, doesn't verify the response' status_code.
+        """
+        url = self.get_url(reverse_args, **kwargs)
+        response = self.url_method(url, **kwargs)
+        if expected_response_code is not None:
+            self.assertEqual(response.status_code, expected_response_code)
+        return response
+
+    def get_url(self, reverse_args=None, **kwargs):  # pylint: disable=unused-argument
+        """Base implementation that returns URL for endpoint that's being tested."""
+        return reverse(self.REVERSE_NAME, kwargs=reverse_args)  # pylint: disable=no-member
+
+    def url_method(self, url, **kwargs):  # pylint: disable=unused-argument
+        """Base implementation that returns response from the GET method of the URL."""
+        return self.client.get(url)
+
+
+class MobileCourseAPITestCase(MobileAPITestCase):
```

Comment:
I found the proliferation of mixins with a deep class hierarchy kind of confusing here and it makes it harder to just compose things.

My naive thought would be that something that had a course and a user argument would mix in the course test case and the user test case, but how it is here, they both bring in a whole bunch of (the same) functionality and it requires having a third custom combination. I imagine we'll have a few more types that are super common and having to have a mixin for each combination seems excessive.

As a proposal, what do you think of the idea of making ``reverse_args`` a method that can be called with decorators that add to its output.

So you wouldn't need e.g. MobileUserCourseAPITestCase. There would just be just a ``user_argument`` class decorator and a ``course_argument`` class decorator. Each of these would append the relevant thing to the result of a ``get_reverse_args`` method that as a stub returns an empty dict.

For extra convenience if we wanted it, we could have MobileUserCourseAPITestCase then be the class that inherits from ``MobileAPITestCase`` with both of those decorators.